# 1. Research Question

## Research Question

Can we guess a song's musical genre just from its lyrics?

A country ballad and a hip-hop track feel completely different on the page — but is that intuition backed by measurable linguistic patterns? We frame this as a **multi-class text classification** problem: given only raw lyrics, predict whether a song is Rock, Pop, Hip-Hop, Metal, Country, Jazz, Electronic, Folk, R&B, or Indie.

- **Task:** Multi-class text classification.
- **Input:** Song lyrics (raw text).
- **Target variable:** Musical genre (categorical label).
- **Hypothesis:** Each genre leaves a distinct linguistic fingerprint — in vocabulary, syntax, and rhetoric — strong enough to beat a majority-class baseline.

**Expected observations:** Hip-Hop stands out through slang and self-reference; Country through rural imagery; Metal through aggressive vocabulary. Syntactically, Hip-Hop favors short lines while Folk leans narrative. Models that capture context and word order should outperform simpler bag-of-words approaches.

---

## Required Reflection

### 1. Why is this question non-trivial?

At first glance, guessing genre from lyrics sounds straightforward — but it really isn't, for a few reasons:

- **Genre boundaries are blurry.** A song tagged as Pop-Rock on Spotify doesn't neatly fit either category. Labels are often assigned by platforms or marketing, not by any linguistic criterion.
- **We're throwing away most of the signal.** Without melody, tempo, or instrumentation, the classifier only gets the text. That's a fraction of what actually defines a genre.
- **Huge variation within genres.** A quiet acoustic Rock ballad and an arena Rock anthem share a label but almost nothing else in their lyrics.
- **Class imbalance.** Rock dominates most lyrics datasets, which means the model could learn to just predict "Rock" and still look decent on accuracy.

### 2. What linguistic phenomenon does it involve?

At its core, this project is about **register variation conditioned by genre** — the idea that the social and artistic context of a song shapes how language is used. More concretely, we're looking at:

- **Lexical semantics:** Do genres prefer different words and word distributions?
- **Morpho-syntactic patterns:** Are there measurable differences in POS-tag usage, sentence length, or syntactic complexity across genres?
- **Pragmatics and rhetoric:** Do genres differ in their use of repetition, imperatives, interjections, or figurative language?
- **Sociolinguistic register:** How much do formality, slang density, and code-switching vary from one genre to another?

### 3. What would falsify our hypothesis?

We'd have to admit we were wrong if:

- All models — including the most expressive ones — perform around the majority-class baseline, which would mean lyrics simply don't carry enough genre signal.
- Our linguistic analyses show no statistically significant vocabulary, syntactic, or rhetorical differences across genres.
- Misclassification errors are scattered randomly instead of clustering around genre pairs that are linguistically close (e.g., Rock/Metal, Folk/Country), which would suggest the models aren't really learning meaningful language patterns.

# 2. Data & Preprocessing

## Loading the dataset

In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("akyshnik/lyrics")

print("Path to dataset files:", path)

Path to dataset files: /home/alberto/.cache/kagglehub/datasets/akyshnik/lyrics/versions/1


In [2]:
import os

print(os.listdir(path))

['kaggle_clean.csv']


In [3]:
import pandas as pd

file_path = os.path.join(path, "kaggle_clean.csv")
df = pd.read_csv(file_path)

## 2.1 Dataset Selection

### Checking the values

In [4]:
print(df["genre"].unique())
df.info(),
df.sample(10)

['Pop' 'Hip-Hop' 'Not Available' 'Rock' 'Metal' 'Other' 'Country' 'Jazz'
 'Electronic' 'Folk' 'R&B' 'Indie']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266557 entries, 0 to 266556
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  266557 non-null  int64 
 1   index       266557 non-null  int64 
 2   song        266556 non-null  object
 3   year        266557 non-null  int64 
 4   artist      266557 non-null  object
 5   genre       266557 non-null  object
 6   lyrics      266535 non-null  object
 7   wnum        266557 non-null  int64 
dtypes: int64(4), object(4)
memory usage: 16.3+ MB


,Unnamed: 0,index,song,year,artist,genre,lyrics,wnum
23566,23566,32937,really-don-t-care-solo-version,2016,demi-lovato,Pop,you wanna play you wanna stay you wanna have i...,440
79148,79148,109859,life-is-just-a-bowl-of-cherries,2011,doris-day,Jazz,people are queer they're always crowing scramb...,197
94402,94402,129464,longer-than-always,2007,dolly-parton,Country,is forever longer than always dolly parton lyr...,148
201676,201676,274541,sabertooth-tiger,2011,cage-the-elephant,Rock,on the other side of town lives a sabertooth t...,250
119507,119507,163055,revolution-without-arms,2007,amaran,Metal,john doe the answer to my prayers\nwon't you c...,65
10542,10542,14817,i-do,2004,anastacia,Rock,who are they anyway\nwith their sticks and sto...,210
231884,231884,314792,take-a-load-off-your-feet,2006,beach-boys,Rock,i do them when i'm down in the tub\nwith avoca...,240
165302,165302,225356,daddy-dearest,2007,dogwood,Rock,what did she do to you sir?\ncan't you see tha...,297
237254,237254,322097,folks-who-live-on-the-hill,2005,bette-midler,Rock,many men with lofty aims strive for lofty goal...,150
109786,109786,150000,louise,2014,erroll-garner,Other,peg o'my heart\ni love you don't let us part\n...,54


In [5]:
print(len(df))
df = df.dropna(subset=["lyrics"])
print(len(df))

266557
266535


### Drop columns with useless information

In [6]:
df = df.drop(columns=["Unnamed: 0", "index"])

### Modify useless values

We modify every value from the lyrics that are formed by one word to be "instrumental"

In [7]:
df.loc[df["wnum"] == 1, "lyrics"].value_counts()

lyrics
instrumental     3645
instru             51
instumental        10
intro              10
intrumental         9
                 ... 
orchestra           1
finntroll           1
pow!                1
instermentual       1
bonus               1
Name: count, Length: 73, dtype: int64

In [8]:
df.loc[df["wnum"] == 1, "lyrics"] = "instrumental"
df.loc[df["wnum"] == 1, "lyrics"].value_counts()

lyrics
instrumental    3820
Name: count, dtype: int64

### Eliminate Non Available values

We do this because there is enough values to being able to eliminate every row that has incomplete information.

In [9]:
df_clean = df[~df["genre"].isin(["Not Available"])]
df_clean = df_clean[~df_clean["year"].isin([112, 702, 67])]

In [10]:
print(len(df), len(df_clean))

266535 242595


### Check impact of instrumental songs

In [11]:
print("Percentage of instrumental songs for each genre:\n")

inst = df_clean["genre"][df_clean["lyrics"] == "instrumental"].value_counts()
no_inst = df_clean["genre"][df_clean["lyrics"] != "instrumental"].value_counts()
genres = inst.keys()

for i in range(len(inst)):
    print(genres[i]+": ", round((inst.iloc[i]/no_inst.iloc[i])*100, 2), "\b%")

Percentage of instrumental songs for each genre:

Rock:  1.3%
Metal:  1.9%
Electronic:  1.4%
Pop:  0.88%
Jazz:  1.01%
Folk:  1.73%
Hip-Hop:  1.1%
Country:  1.23%
Indie:  1.33%
R&B:  0.32%
Other:  0.28%


### Check number of values of each genre

In [12]:
genre_values = df_clean["genre"].value_counts()
df_size = len(df_clean)
biggest_value = 0
for genre in genres:
    if len(genre) > biggest_value:
        biggest_value = len(genre)
for i in range(len(genres)):
    rounded_number = round((genre_values.iloc[i]/df_size)*100, 2)
    print(genres[i]+":"+" "*(biggest_value-len(genres[i]) + 5-len(str(rounded_number))), 
          rounded_number, "\b%\t", genre_values.iloc[i])

Rock:       45.02%	 109221
Metal:      16.68%	 40465
Electronic: 10.24%	 24844
Pop:         9.79%	 23759
Jazz:        5.93%	 14387
Folk:        3.29%	 7971
Hip-Hop:     3.28%	 7966
Country:     2.14%	 5189
Indie:        1.4%	 3401
R&B:          1.3%	 3149
Other:       0.92%	 2243


### Dataset Description

- Source: https://www.kaggle.com/datasets/akyshnik/lyrics
- Domain: Prediction of music genre from the lyrics.
- Label distribution:
| Genre | Percentage | Samples |
| --- | --- | --- |
| Rock | 54.02% | 109235 |
| Metal | 16.68% | 40465 |
| Electronic | 10.24% | 24845 |
| Pop | 9.79% | 23759 |
| Jazz | 5.93% | 14387 |
| Folk | 3.29% | 7971 |
| Hip-Hop | 3.28% | 7966 |
| Country | 2.14% | 5189 |
| Indie | 1.4% | 3401 |
| R&B | 1.3% | 3149 |
| Other | 0.92% | 2243 |
- Potential biases: The dataset can be biased towards Rock music, since almost half of the music from this dataset is Rock. But, even with this statement, the most represented genres (Rock, Metal, Electronic, Pop) have subgenres inside them (blues rock, goth metal, afro music, pop-punk). So it makes sense for these genres to have the most amount of samples.
- Ethical considerations: All the information of the dataset is public. Some songs contain offensive language, but this shouldn't be taken as harmful content, since offensive language also helps defining a way of expressing, thus helping on the genre prediction.

## 2.2 Data Split

### Data Split justification

For the data split, we will give 70% to the train dataset, 15% to the development dataset, and 15% to the test dataset.
The reason of this division is to train the dataset with the most amount of data, so the model can learn patterns correctly. For this reason, the development and test datasets should be big enough to work properly, but as small as possible to use the most amount of data on the training dataset.

The randomness is controlled via the random_state parameter. By applying a value on this parameter, we ensure that every time we repeat the executions, the dataset will split in the same way.

We also use stratification to keep the label proportions evenly in each dataset, since it wouldn't be ideal to have one of the datasets almost full of Rock songs, for example.

In [13]:
from sklearn.model_selection import train_test_split

df_train, df_temp = train_test_split(df_clean, test_size=0.3, stratify=df_clean["genre"], random_state=42)
df_dev, df_test = train_test_split(df_temp, test_size=0.5, stratify=df_temp["genre"], random_state=42)

In [14]:
print(len(df_clean), len(df_train), len(df_dev), len(df_test))

242595 169816 36389 36390


## 2.3 Preprocessing strategy

In [15]:
df_train.info()
df_train.sample(10)

<class 'pandas.core.frame.DataFrame'>
Index: 169816 entries, 229632 to 68966
Data columns (total 6 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   song    169815 non-null  object
 1   year    169816 non-null  int64 
 2   artist  169816 non-null  object
 3   genre   169816 non-null  object
 4   lyrics  169816 non-null  object
 5   wnum    169816 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 9.1+ MB


,song,year,artist,genre,lyrics,wnum
212859,lovely-rita,2009,cheap-trick,Rock,lovely rita meter maid\nlovely rita meter maid...,184
73699,good-man,2013,devour-the-day,Rock,i want to be a good man i want to see god\ni w...,288
217365,bed-spins,2014,the-dying-arts,R&B,bed spins\nshe'll close her mouth and nothing'...,158
54782,jet-black-sunrise,2007,gin-blossoms,Rock,time passes like a dream\nshe suspends the wor...,216
5864,richard-hung-himself,2007,d-i,Rock,needles were stabbed into the walls\nthe execu...,287
33702,though-the-days-were-only-seven,2007,ernest-tubb,Country,angels sing sweet songs of love dear gates of ...,135
170930,if-we-try,2006,babyface,R&B,we go through changes\ntime after time\ni gues...,189
158395,lullaby-of-birdland,2009,charlie-parker,Jazz,lullaby of birdland that's what i\nalways hear...,107
113126,all-sinks-in,1990,connells,Rock,i've been turning a new direction\nand polishi...,294
64195,get-a-life,2006,des-ree,Pop,wake up greet the angels\nsit back and wipe yo...,142


* tokenization method: word-level tokenization, since the meaning of words are very useful to identify the music genre.
* lemmatization vs stemming: lemmatization, because we want to the actual meaning of words, which is also very useful to identify the music genre.
* stopword removal (or not): we will keep them, since there are genres that don't use many and will be easier to identify.
* rare word filtering: we will take them out, because rare words might increase psarsity.
* handling of emojis / punctuation: there are no emojis. We will keep the punctuation in case we use transformers.
* subword tokenization (if transformer): we will use a pre-trained transformer, such as BERT.

In [16]:
# Unique words (Vocabulary Size)
# We convert to lowercase to avoid counting "The" and "the" as different words
initial_vocab = set()
df_train['lyrics'].str.lower().str.split().apply(initial_vocab.update)

# Average Words per Song
# split() counts every word separated by whitespace
avg_words_before = df_train['lyrics'].apply(lambda x: len(x.split())).mean()

print(f"Vocab Size Before: {len(initial_vocab)}")
print(f"Avg Words per Song Before: {avg_words_before:.2f}")

Vocab Size Before: 453369
Avg Words per Song Before: 228.45


### Preprocessing application

In [17]:
import re
from collections import Counter

# 1. Define a standard tokenization function (as per teacher's guide)
def tokenize(text):
    return re.findall(r"\b\w+\b", text.lower())

# 2. Applying tokenization to the dataset
df_train['tokens'] = df_train['lyrics'].apply(tokenize)

# 3. Handling Stopwords
with open('EN-Stopwords.txt', 'r', encoding='utf-8') as f:
    stopwords = set(line.strip().lower() for line in f if line.strip())
df_train['tokens'] = df_train['tokens'].apply(lambda x: [t for t in x if t not in stopwords])

# 4. Rare Word Filtering
vocab_counter = Counter()
df_train['tokens'].apply(vocab_counter.update)

# We identify words that appear fewer than 5 times
min_freq = 5
rare_words = {word for word, count in vocab_counter.items() if count < min_freq}

# Filtering of rare words
df_train['tokens'] = df_train['tokens'].apply(lambda x: [t for t in x if t not in rare_words])

In [18]:
from collections import Counter

# 1. Vocabulary Size After
final_vocab = Counter()
df_train['tokens'].apply(final_vocab.update)

# 2. Average Tokens per Song After
avg_tokens_after = df_train['tokens'].apply(len).mean()

print(f"Vocab Size After: {len(final_vocab)}")
print(f"Avg Tokens per Song After: {avg_tokens_after:.2f}")

Vocab Size After: 87004
Avg Tokens per Song After: 76.85
